Add models, validations and associations

katemildner opened the pull request:
Created models:
- CinemaHall
- Movie
- Reservation
- Role
- Screening
- SeatReservation
- Seat
- Status

Added validations and associations for models.
Fixed offenses raised by rubocop.

katemildner committed:
change sqlite3 for postgresql as the database

katemildner committed:
configure postgresql database

katemildner committed:
[fix] Change faulty configuration in database.yml

Couldn't run migration due to the "ActiveRecord::ConnectionNotEstablished: fe_sendauth: no password supplied" error, changing username fixed connection with db.

katemildner committed:
Add CinemaHall model and corresponding migration

katemildner committed:
Add options null:false for CinemaHall migration; add current schema

katemildner committed:
Add validations to CinemaHall model

Added presence validations for all attributes and numericality
validations for integer attributes - this will provide better
error messages than errors directly from db.

katemildner committed:
Add Movie model and corresponding migration

katemildner committed:
Add options null:false for movies migration; add current schema

katemildner committed:
Add validations to Movie model

Added presence validations for all attributes - this will provide
better error messages than errors directly from db.

katemildner committed:
Add Screening model and corresponding migration

katemildner committed:
Add options null:false for screenings migration; add current schema

katemildner committed:
Add validations to Screening model

Added presence validation for the attribute - this will provide
better error messages than errors directly from db.

katemildner committed:
Add Seat model and corresponding migration

katemildner committed:
Add options null:false for seats migration; add current schema

katemildner committed:
Add validations to Seat model

Added presence validations for all attributes - this will provide
better error messages than errors directly from db.

katemildner committed:
Add Status model and corresponding migration

katemildner committed:
Add options null:false for statuses migration; add current schema

katemildner committed:
Add validations to Status model

Added presence validations for all attributes - this will provide
better error messages than errors directly from db.

katemildner committed:
Add Reservation model, corresponding migration and updated schema

katemildner committed:
Add SeatReservation model, corresponding migration and updated schema

katemildner committed:
Add Role model and corresponding migration

katemildner committed:
Add options null:false for roles migration; add current schema

katemildner committed:
Add validations to Role model

Added presence validations for all attributes - this will provide
better error messages than errors directly from db.

katemildner committed:
[refactor] Fix style offenses in all models

katemildner committed:
[refactor] Fix style offenses in migrations

Excluded schema.db entirely in rubocop config, excluded migrations for Documentation cop.

katemildner committed:
Add associations to models

katemildner committed:
[fix] Change associations between reservation and status

katemildner committed:
Specified :dependent option for has_many associations

kodowicz commented on db/migrate/20210819170858_create_screenings.rb:
```diff
@@ -0,0 +1,11 @@
+class CreateScreenings < ActiveRecord::Migration[6.1]
+  def change
+    create_table :screenings do |t|
+      t.datetime :date_and_time, null: false
```
Maybe more readable would be `starts_at` instead of `date_and_time`?:)

katemildner replied:
I wanted this to explicitly state screening requires starting date and time, but now I see how unclear this may be. I'll come up with something more descriptive, i.e. _starting_ _date_and_time. :)

katemildner committed:
[style] Change column date_and_time name in screenings to more explicit

For the sake of clear naming changed date_and_time to starting_date_and_time.

w0st commented on db/migrate/20210819170858_create_screenings.rb:
```diff
@@ -0,0 +1,11 @@
+  def change
+    create_table :screenings do |t|
+      t.datetime :date_and_time, null: false
+
```
Looking at [the model](https://github.com/katemildner/Monte-Cinema/pull/4/files#diff-9985e652ede97744897222361bffd26992506744b6b06a36ca22f6051a90640aR6) it seems that this migration should have a reference to `cinema_hallls` relation, but haven't.

Read about a [reference](https://edgeguides.rubyonrails.org/active_record_migrations.html#references) in migrations which will create a ForeignKey to `cinema_halls` table, please.

https://riptutorial.com/ruby-on-rails/example/3492/add-a-reference-column-to-a-table
https://api.rubyonrails.org/v5.1.7/classes/ActiveRecord/ConnectionAdapters/TableDefinition.html#method-i-references


The same with a FK to `movies`

katemildner replied:
Thanks, I remember now: I planned to add FKs after all my models will be implemented and this will happen after adding devise and creating User model, but I agree - there's no need to postpone adding Foreign Keys for already existing associations.

Therefore: Foreign Keys are added.

w0st commented on db/migrate/20210819172106_create_seats.rb:
```diff
@@ -0,0 +1,12 @@
+
+class CreateSeats < ActiveRecord::Migration[6.1]
+  def change
+    create_table :seats do |t|
```
Consider adding a FK to `cinema_halls`

w0st commented on db/migrate/20210819182409_create_reservations.rb:
```diff
@@ -0,0 +1,7 @@
+
+class CreateReservations < ActiveRecord::Migration[6.1]
+  def change
+    create_table :reservations, &:timestamps
```
Add FK to `statuses`, `screenings`, please.

w0st commented on db/migrate/20210819192713_create_seat_reservations.rb:
```diff
@@ -0,0 +1,7 @@
+
+class CreateSeatReservations < ActiveRecord::Migration[6.1]
+  def change
+    create_table :seat_reservations, &:timestamps
```
Add FK to `reservations`, `seats`, please.

w0st commented on app/models/movie.rb:
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+class Movie < ApplicationRecord
+  validates :title, presence: true
```
You can pass to [`validates`](https://api.rubyonrails.org/v5.1.7/classes/ActiveModel/Validations/ClassMethods.html#method-i-validates) more than one attribute. So you can write it in that way: `validates :title, :description, presence: true` etc.

Feel free to ignore this comment, it's just a hint. The final decision is up to you and your personal style.

katemildner replied:
Thank you, I'll keep it that way for readability's sake. I know there's not that much attributes here to validate, but I'm just exercising good habits here. ;)

w0st replied:
It's fine :ok_hand:

w0st reviewed (changes requested):
The migrations don't contains foreign keys where you most likely would like to have them :wink:

katemildner committed:
Add foreign keys to the entire schema

Added all required FKs; changed the name of column date_and_time to screening_date_and_time in the screening table and removed separate migration for this change for better clarity.

w0st reviewed (approved):
It looks good to me. Good job!
